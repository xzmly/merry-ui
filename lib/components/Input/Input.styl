@import '../../helpers/stylus.helper.styl';

label = class-prefix + "-input-label"
input = class-prefix + "-input"

default-size-height = 32px
small-size-height = 24px
big-size-height = 40px

common-size-style(padding,fontSize,height)
  padding padding
  font-size fontSize
  height height


.{label}
  font-size min-font-size
  vertical-align middle
  color font-default-color

.{input}
  outline none
  border 1px solid #d9d9d9
  height default-size-height
  background #fff
  padding 4px 11px
  box-sizing border-box
  font-size min-font-size
  border-radius 4px
  color font-default-color
  display inline-block
  line-height 1.5
  transition all .2s
  &.{input}-small
    common-size-style(1px 7px,very-min-font-size,small-size-height)
  &.{input}-big
    common-size-style(6px 11px,middle-font-size,big-size-height)
  &:focus
    outline none
    border-color warning-color
    box-shadow: 0 0 0 2px rgba(255, 185, 17, 0.23)
