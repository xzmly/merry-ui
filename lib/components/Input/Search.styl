@import "./Input.styl"

search = input + "-search"


.{search}
  .{input}-affix-wrap
    display flex
    .{input}
      border-radius: 4px 0 0 4px
      &:not(:last-child)
        padding-right: 11px;
    .{input}-affix-suffix
      position relative
      right 0
      top 0
      transform none
      .search-btn
        min-width 46px
        margin 0
        border-radius 0 4px 4px 0
        &.merry-btn-big
          height 40px

