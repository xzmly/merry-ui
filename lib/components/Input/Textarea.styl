@import '../../helpers/stylus.helper.styl';

textarea = class-prefix + "-textarea"

hover-and-focus()
  transition all .2s,height 0s
  outline none
  border-color warning-color
  box-shadow: 0 0 0 2px rgba(255, 185, 17, 0.23)

.{textarea}
  display flex
  > textarea
    line-height 1.5
    font-size min-font-size
    outline none
    border 1px solid #d9d9d9
    border-radius 4px
    box-sizing border-box
    font-family: inherit;
    color font-default-color
    flex 1
    display inline-block
    background-color #fff
    background-image none
    &:focus
      hover-and-focus()
    &:hover
      hover-and-focus()
  > span
    color font-default-color
    &:after
      content ':'
      padding 0 6px 0 3px
      top -0.6px
      position relative

.{textarea}-top
  flex-direction column
  > textarea
    max-width 100%
    min-width 100%
    flex none
  > span
    padding-bottom 8px
    &:after
      display none
