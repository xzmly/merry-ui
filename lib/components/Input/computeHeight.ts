const hidden_style = `
  min-height:0 !important;
  max-height:none !important;
  height:0 !important;
  visibility: hidden !important,
  overflow: hidden !important,
  position: absolute !important,
  zIndex: -1000 !important,
  top: -99px !important,
  right: -999px !important,
  opacity: 0 !important
`;

const SIZING_STYLE = [
  'letter-spacing',
  'line-height',
  'padding-top',
  'padding-bottom',
  'font-family',
  'font-weight',
  'font-size',
  'font-variant',
  'text-rendering',
  'text-transform',
  'width',
  'text-indent',
  'padding-left',
  'padding-right',
  'border-width',
  'box-sizing',
];

export interface NodeType {
  sizingStyle: string;
  paddingSize: number;
  borderSize: number;
  boxSizing: string;
}

let hiddenTextarea: HTMLTextAreaElement;
const computedStyleCache: { [key: string]: NodeType } = {};

export function copyTextareaStyle(node: HTMLElement){
  const className:string = node.getAttribute('name') as string;

  if(computedStyleCache[className]){
    return computedStyleCache[className]
  }

  const style = window.getComputedStyle(node);

  const boxSizing:string =
      style.getPropertyValue('box-sizing') ||
      style.getPropertyValue('-moz-box-sizing') ||
      style.getPropertyValue('-webkit-box-sizing');

  const paddingSize:number =
      parseFloat(style.getPropertyValue('padding-bottom')) +
      parseFloat(style.getPropertyValue('padding-top'));

  const borderSize:number =
      parseFloat(style.getPropertyValue('border-bottom-width')) +
      parseFloat(style.getPropertyValue('border-top-width'));

  const sizingStyle = SIZING_STYLE.map(name => `${name}:${style.getPropertyValue(name)}`).join(';');

  const nodeInfo: NodeType = {
    sizingStyle,
    paddingSize,
    borderSize,
    boxSizing,
  };

  if (!computedStyleCache[className]) {
    computedStyleCache[className] = nodeInfo;
  }

  return nodeInfo;

}

export function computeHeight(
    node: HTMLTextAreaElement,
    minRows: number | null = null,
    maxRows: number | null = null
){
  if(!hiddenTextarea) {
    hiddenTextarea = document.createElement('textarea');
    hiddenTextarea.setAttribute('style',hidden_style);
    document.body.appendChild(hiddenTextarea)
  }

  console.log(minRows,maxRows);

  const { paddingSize, borderSize, boxSizing, sizingStyle } = copyTextareaStyle(node);

  hiddenTextarea.setAttribute('style',`${sizingStyle};${hidden_style}`)
  hiddenTextarea.value = node.value || node.placeholder || '';

  let minHeight = Number.MIN_SAFE_INTEGER;
  let maxHeight = Number.MAX_SAFE_INTEGER;
  let height = hiddenTextarea.scrollHeight;
  let overflowY: any;

  if (boxSizing === 'border-box') {
    // border-box: add border, since height = content + padding + border
    height += borderSize;
  } else if (boxSizing === 'content-box') {
    // remove padding, since height = content
    height -= paddingSize;
  }

  if (minRows !== null || maxRows !== null) {
    // measure height of a textarea with a single row
    hiddenTextarea.value = ' ';
    const singleRowHeight = hiddenTextarea.scrollHeight - paddingSize;
    if (minRows !== null) {
      minHeight = singleRowHeight * minRows;
      if (boxSizing === 'border-box') {
        minHeight = minHeight + paddingSize + borderSize;
      }
      height = Math.max(minHeight, height);
    }
    if (maxRows !== null) {
      maxHeight = singleRowHeight * maxRows;
      if (boxSizing === 'border-box') {
        maxHeight = maxHeight + paddingSize + borderSize;
      }
      overflowY = height > maxHeight ? '' : 'hidden';
      height = Math.min(maxHeight, height);
    }
  }

  return { height, minHeight, maxHeight, overflowY };
}