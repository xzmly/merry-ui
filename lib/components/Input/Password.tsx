import * as React from 'react';
import { useState } from "react"
import Input,{TotalProps}from "./Input";
import Icon from "../Icon/Icon"

const Password: React.FC<TotalProps> =
        props => {

  const [visible,setVisible] = useState<Boolean>(false);

  const { children,...restProps} = props;

  return (
      <Input  {...restProps}
              type={!visible ? 'password':'text'}
              suffix={
                <Icon name={visible ? 'eye' : 'close_eye'}
                      onClick={()=>setVisible(!visible)}
                />
              }>
        {children}
      </Input>
  )
};

export default Password