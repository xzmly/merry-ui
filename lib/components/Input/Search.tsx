import * as React from 'react';
import {useState,useEffect} from "react";
import Input,{TotalProps}from "./Input";
import classes from '../../helpers/classes';
import Icon from "../Icon/Icon"
import Button from "../Button/Button"
import "./Search.styl"

type DefaultValueType = string | number | string[] | undefined

type onChangeType = (event:React.ChangeEvent<HTMLInputElement>) => void

export interface SearchProps extends TotalProps{
  onSearch?: (value: DefaultValueType,e:React.MouseEvent) => void
  enterButton?: boolean | React.ReactNode
}

const Search: React.FC<SearchProps> =
  props => {

    const {className,children,onSearch,enterButton,...restProps} = props;

    const [defaultValue,setDefaultValue] = useState<DefaultValueType>(restProps?.defaultValue);

    useEffect(()=>{
      setDefaultValue(restProps?.defaultValue)
    },[props.defaultValue]);

    const _onSearch = (event:React.MouseEvent):void => {
      onSearch && onSearch(restProps.value || defaultValue,event)
    };

    const onChange:onChangeType = (e) => {
      restProps.onChange ? restProps.onChange(e) : setDefaultValue(e.target.value)
    };

  return (
      <Input {...restProps}
             className={classes(className,'input-search')}
             onChange={onChange}
             defaultValue={defaultValue}
             suffix={
               enterButton ?
                   <Button className={'search-btn'} onClick={_onSearch} size={restProps.size}>
                     {
                       typeof enterButton === 'boolean' ?
                         <Icon name={'search'}/> :
                         enterButton
                     }
                   </Button> :
                   <Icon name={'search'} onClick={_onSearch}/>
             }>
        {children}
      </Input>
  )
};

export default Search