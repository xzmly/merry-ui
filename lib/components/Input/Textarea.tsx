import * as React from 'react';
import { useState,useEffect } from "react"
import { LabelProps } from "./Input";
import classes from '../../helpers/classes';
import { computeHeight } from "./computeHeight";
import "./Textarea.styl";

export interface AutoSizeType {
  minRows?: number,
  maxRows?: number
}

export interface TextareaProps extends React.TextareaHTMLAttributes<HTMLTextAreaElement>{
  autoSize?: AutoSizeType | boolean
  labelPosition?: "top" | "left"
}

export type onChangeType = (event:React.ChangeEvent<HTMLTextAreaElement>) => void

const Textarea: React.FC<TextareaProps & LabelProps> =
    props => {

      const [textareaStyle,setTextareaStyle] = useState<any>({});
      const ref = React.createRef<HTMLTextAreaElement>();

      const {
        children,
        className,
        htmlFor,
        form,
        style,
        labelPosition,
        autoSize,
        ...textareaProps
      } = props;

      useEffect(()=>{
        updateStyle()
      },[]);

      const updateStyle = ():void => {
        const style = autoSize ? computeHeight(ref.current as HTMLTextAreaElement,{
          minRows: _autoSize('minRows'),
          maxRows: _autoSize('maxRows')
        }) : {};

        setTextareaStyle(style);
      };

      const labelProps: LabelProps = { htmlFor, form, style };

      const positionClass:string =
          labelPosition === 'left' ? '' : `textarea-${labelPosition}`;

      const _autoSize = (key: string): undefined | number =>
          typeof autoSize === 'boolean' ? undefined : (autoSize as any)[key];


      const onChange:onChangeType = (event) => {

        updateStyle();
        textareaProps.onChange && textareaProps.onChange(event)
      };

      return (
          <label
              {...labelProps}
              className={classes(className,"textarea",positionClass)}
          >
            {children && <span>{children}</span>}
            <textarea ref={ref}
                      {...textareaProps}
                      style={...textareaStyle}
                      onChange={onChange}/>
          </label>
      )
    };


Textarea.defaultProps = {
  labelPosition: "left"
};

export default Textarea