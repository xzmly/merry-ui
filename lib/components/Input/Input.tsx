import * as React from 'react';
import classes from '../../helpers/classes';
import Password,{ PasswordProps } from "./Password";
import Textarea,{ TextareaProps} from "./Textarea";
import "./Input.styl";

interface LabelProps {
  className?: string
  children?: React.ReactNode
  htmlFor?: string
  form?: string
  style?: React.CSSProperties
}

type restProps = {
  size?: "default" | "small" | "big"
  labelPosition?: "top" | "left"
  type?: "text" | "password" | "textarea"
}

type InputProps = Omit<React.InputHTMLAttributes<HTMLInputElement>,"size" | "type">

interface InputComponent<T> extends React.FC<T>{
  Password: React.FC<PasswordProps>
  Textarea: React.FC<TextareaProps>
}

const Input: InputComponent<LabelProps & InputProps & restProps> =
        props => {

  const {
    className,
    htmlFor,
    form,
    style,
    children,
    size,
    labelPosition,
    type,
    ...inputProps
  } = props;

  const labelProps: LabelProps = { htmlFor, form, style };

  const sizeClass: string =
      size === "default" ? "" : `input-${size}`;

  const labelPsClass: string =
      labelPosition === "left" ? "" : `input-label-${labelPosition}`;

  return (
      <label className={classes(className,"input-label",labelPsClass)}
             {...labelProps}>
        {
          children &&
          <span className={classes("","input-label-text")}>
            {children}
          </span>
        }
        <input {...inputProps}
               className={classes('',"input",sizeClass)}
               type={type}/>
      </label>
  )
};

Input.defaultProps = {
  size: "default",
  type: "text",
  labelPosition: "left"
};

Input.Password = Password;
Input.Textarea = Textarea;

export default Input