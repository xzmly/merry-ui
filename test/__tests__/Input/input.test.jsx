import renderer from 'react-test-renderer'
import Input from '../../../lib/components/Input/Input'
import React from 'react'
import {mount} from "enzyme/build/index";


describe('input',()=>{
  it('render input',()=>{
    const json = renderer.create(<Input/>).toJSON();
    expect(json).toMatchSnapshot()
  })
});