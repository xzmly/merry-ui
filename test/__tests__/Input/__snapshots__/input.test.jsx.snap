// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input render input 1`] = `
<label
  className="merry-input-label"
>
  <input
    className="merry-input"
    type="text"
  />
</label>
`;
